implement transport api for PPL inter-plugin communication

zhongnansu opened the pull request:
Signed-off-by: Zhongnan Su <szhongna@amazon.com>

### Update
After discussion in the thread here https://github.com/opensearch-project/common-utils/pull/155, we decided for now we'll  keep the transport interfaces in sql repo, instead of common-utils for the following reasons.
1. there's no consumer plugin for this feature yet. Because the response format is not encapsulated. Currently on trasnport layer there's only a json string in the response. Ideally we want a `ResultSet` or `DataFrame` like response for clients to access data. An issued is created w.r.t this.
2. This change is to the request model, which is the input of the SQL/PPL engine. Having it separated into 2 repos is not development friendly. E.g, the CI will fail for sure, because CI only pulls against released `common-utils` 

### Description
Add a transport layer for handle PPL request. This will expose transport interface for inter-plugin communication. The existing Rest layer handler is also being refactored to make use of the transport layer

- Create `TransportPPLQueryAction`
- refactored `RestPPLQueryAction` to service as a route layer, moved the query request handling logic to `TransportPPLQueryAction`.
- Register action and handler in SQLPlugin
- Refactored `PPLQueryRequest` to be writable, for the transport layer commuication
- Create transport response model that currently holds a string response.

- ~~Add common-utils as dependency, since the request/response models will be exposed in common-utils for other plugin to consume as well.~~ 

- **TODO:** 
  1. Request and response format are strings for now.(may need modification in future version, especially response format)
  2. Implement transport interface for SQL, in a similar manner


### Issues Resolved
#521 


### CI/CD
- ~~The CI is failing because of the [pending PR in common-utils](https://github.com/opensearch-project/common-utils/pull/155)~~
- [Fixed]~~local build failed because of jarhell caused by ml-commons. Raised another PR in their repo to fix. https://github.com/opensearch-project/ml-commons/pull/328~~
```

| java.home: /Library/Java/JavaVirtualMachines/jdk-14.0.2.jdk/Contents/Home
| examining directory: /Users/szhongna/Desktop/Projects/OpenSearch/sql/integ-test/build/classes/java/test
| examining directory: /Users/szhongna/Desktop/Projects/OpenSearch/sql/integ-test/build/resources/test
| examining jar: /Users/szhongna/Desktop/Projects/OpenSearch/sql/sql-jdbc/build/libs/opensearch-sql-jdbc-2.0.0.0.jar
| examining jar: /Users/szhongna/Desktop/Projects/OpenSearch/sql/sql-jdbc/build/libs/opensearch-sql-jdbc-1.2.0.0.jar
| Exception in thread "main" java.lang.IllegalStateException: jar hell!
| class: org.opensearch.jdbc.ResultSetImpl
| jar1: /Users/szhongna/Desktop/Projects/OpenSearch/sql/sql-jdbc/build/libs/opensearch-sql-jdbc-2.0.0.0.jar
| jar2: /Users/szhongna/Desktop/Projects/OpenSearch/sql/sql-jdbc/build/libs/opensearch-sql-jdbc-1.2.0.0.jar
|       at org.opensearch.bootstrap.JarHell.checkClass(JarHell.java:317)
|       at org.opensearch.bootstrap.JarHell.checkJarHell(JarHell.java:212)
|       at org.opensearch.bootstrap.JarHell.checkJarHell(JarHell.java:99)
|       at org.opensearch.bootstrap.JarHell.main(JarHell.java:83)

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':integ-test:jarHell'.
> Process 'command '/Library/Java/JavaVirtualMachines/jdk-14.0.2.jdk/Contents/Home/bin/java'' finished with non-zero exit value 1

```
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

joshuali925 commented on plugin/src/main/java/org/opensearch/sql/plugin/transport/PPLQueryHelper.java:
```diff
@@ -0,0 +1,119 @@
+            INSTANCE = new PPLQueryHelper();
+        }
+        return INSTANCE;
+    }
```
also set constructor to private for singleton?

zhongnansu replied:
good pointed, missed that

joshuali925 commented on plugin/src/main/java/org/opensearch/sql/plugin/transport/PPLQueryHelper.java:
```diff
@@ -0,0 +1,119 @@
+    private static ResponseListener<ExecutionEngine.QueryResponse> createListener(
+            PPLQueryRequest pplRequest,
+            ActionListener<TransportQueryResponse> listener
+    ) {
```
is it possible to reuse/extend RestPPLQueryAction rather than redefining some methods? or is implementation slightly different?

penghuo commented on plugin/src/main/java/org/opensearch/sql/plugin/transport/TransportQueryAction.java:
```diff
@@ -0,0 +1,66 @@
+/**
+ * Send SQL/PPL query transport action
+ */
+public class TransportQueryAction extends HandledTransportAction<ActionRequest, TransportQueryResponse> {
```
should we 
1. create TransportSQLService and TransportPPLService
2. simply RestXXXAction

zhongnansu replied:
good point, renames classed and updated

penghuo reviewed (commented):
seems missing files, SQLActions, TransportQueryResponse

zhongnansu commented:
> seems missing files, SQLActions, TransportQueryResponse

@penghuo Those are interfaces that should be defined in common-utils and used by any caller plugins. Currently I have a [PR](https://github.com/opensearch-project/common-utils/pull/155/files#diff-ffe15eb31a89084a5bef761be8e4af6a9e28aa27b669f874b5bd9ff8ddd49c03) pending for common-utils. Once that's merged and released to maven central, this PR will pass CI

zhongnansu commented on plugin/build.gradle:
```diff
@@ -107,12 +117,14 @@ dependencyLicenses.enabled = false
 testingConventions.enabled = false
 
-// TODO: need to verify the thirdPartyAudi
+// TODO: need to verify the thirdParty
```
I think so, fixed

zhongnansu commented on build.gradle:
```diff
@@ -7,6 +7,9 @@
         opensearch_version = System.getProperty("opensearch.version", "2.0.0-alpha1-SNAPSHOT")
+        opensearch_build = opensearch_version.replaceAll(/([\d.]*\.\d+(?!\.\d))/, '$1.0')
+        common_utils_version = System.getProperty("common_utils.version", opensearch_build)
+        kotlin_version = System.getProperty("kotlin.version", "1.4.32")
```
Because we'll consume the req/res models and interfaces, which are defined in `comon-utils`.

penghuo commented on plugin/src/main/java/org/opensearch/sql/plugin/SQLPlugin.java:
```diff
@@ -110,12 +151,50 @@ public Collection<Object> createComponents(Client client, ClusterService cluster
+    });
+  }
+
+  private SQLService createSQLService(NodeClient client) {
```
new SQLService instance should be in request level, not Plugin level.

zhongnansu replied:
removed

penghuo commented on plugin/src/main/java/org/opensearch/sql/plugin/transport/TransportSQLService.java:
```diff
@@ -0,0 +1,119 @@
+import static org.opensearch.sql.legacy.plugin.RestSqlAction.QUERY_API_ENDPOINT;
+import static org.opensearch.sql.protocol.response.format.JsonResponseFormatter.Style.PRETTY;
+
+public class TransportSQLService {
```
It seems TransportSQLService is a static class. Should we merge TransportSQLService into TransportSQLQueryAction? My previous comments maybe misleading.

penghuo reviewed (commented):
Two things.
1. RestSQLAction should be refactored to call TransportSQLQueryAction.
2. Is it possible to add UT?

penghuo commented on plugin/src/main/java/org/opensearch/sql/plugin/transport/TransportSQLQueryAction.java:
```diff
@@ -0,0 +1,59 @@
+/**
+ * Send SQL query transport action
+ */
+public class TransportSQLQueryAction extends HandledTransportAction<ActionRequest, TransportSQLQueryResponse> {
```
In PR, https://github.com/opensearch-project/common-utils/pull/155/,  TransportQueryRequest is defined, instead of TransportSQLQueryAction

zhongnansu replied:
@penghuo Hi Peng, sorry I forget to submitted the commit to common-utils after the change, could you take a look again? I separated `TransportQueryRequest` into `TrasnportSQLQueryRequest` and `TransportPPLQueryRequest`  https://github.com/opensearch-project/common-utils/pull/155

penghuo commented on plugin/build.gradle:
```diff
@@ -70,8 +70,24 @@ dependencies {
+    api "org.jetbrains.kotlin:kotlin-stdlib-common:${kotlin_version}"
 }
 
+//shadowJar {
+//    // fix jarhell by relocating packages
+//    relocate 'org.opensearch.commons', 'org.opensearch.sql.plugin.repackage.org.opensearch.commons'
+//    classifier = null
+//}
```
delete it.

zhongnansu replied:
sure

penghuo commented on plugin/src/main/java/org/opensearch/sql/plugin/transport/TransportPPLQueryAction.java:
```diff
@@ -0,0 +1,256 @@
+     */
+    private final Settings pluginSettings;
+
+    private final Supplier<Boolean> pplEnabled;
```
not used?

zhongnansu replied:
removed

penghuo commented on plugin/src/main/java/org/opensearch/sql/plugin/transport/TransportPPLQueryAction.java:
```diff
@@ -0,0 +1,256 @@
+            public void onResponse(ExecutionEngine.QueryResponse response) {
+                String responseContent = formatter.format(new QueryResult(response.getSchema(),
+                        response.getResults()));
+                sendResponse(responseContent, listener);
```
listener.onResponse(new TransportPPLQueryResponse(content));

penghuo commented on plugin/src/main/java/org/opensearch/sql/plugin/transport/TransportPPLQueryAction.java:
```diff
@@ -0,0 +1,256 @@
+
+            @Override
+            public void onFailure(Exception e) {
+                LOG.error("Error happened during query handling", e);
+                if (isClientError(e)) {
+                    Metrics.getInstance().getNumericalMetric(MetricName.PPL_FAILED_REQ_COUNT_CUS).increment();
+                    reportError(listener, e, BAD_REQUEST);
+                } else {
+                    Metrics.getInstance().getNumericalMetric(MetricName.PPL_FAILED_REQ_COUNT_SYS).increment();
+                    reportError(listener, e, SERVICE_UNAVAILABLE);
+                }
```
listener.onFailure(e)

zhongnansu replied:
make sense

penghuo commented on plugin/src/main/java/org/opensearch/sql/plugin/rest/RestPPLQueryAction.java:
```diff
@@ -122,138 +96,34 @@ protected Set<String> responseParams() {
+            null);
+
+    return channel -> nodeClient.execute(SQLActions.INSTANCE.getSEND_PPL_QUERY_ACTION_TYPE(),
+            transportPPLQueryRequest, new RestToXContentListener<>(channel){
```
new ResponseListener<TransportPPLQueryResponse>() {
      @Override
      public void onResponse(QueryResponse response) {
        sendResponse(channel, OK, formatter.format(new QueryResult(response.getSchema(),
            response.getResults())));
      }

      @Override
      public void onFailure(Exception e) {
        LOG.error("Error happened during query handling", e);
        if (isClientError(e)) {
          Metrics.getInstance().getNumericalMetric(MetricName.PPL_FAILED_REQ_COUNT_CUS).increment();
          reportError(channel, e, BAD_REQUEST);
        } else {
          Metrics.getInstance().getNumericalMetric(MetricName.PPL_FAILED_REQ_COUNT_SYS).increment();
          reportError(channel, e, SERVICE_UNAVAILABLE);
        }
      }
}

zhongnansu replied:
sure, but I will leave the formatter in the transport layer, and only pass the formatted response string in `onResponse()`

penghuo commented on plugin/src/main/java/org/opensearch/sql/plugin/transport/TransportSQLQueryAction.java:
```diff
@@ -0,0 +1,60 @@
+ * TODO: we'll refactor this the similar way as TransportPPLAction
+ * Send SQL query transport action
+ */
+public class TransportSQLQueryAction extends HandledTransportAction<ActionRequest, TransportSQLQueryResponse> {
```
Remove from this PR if it is not used.

zhongnansu replied:
removed

zhongnansu commented:
Addressed all comments. Local build keeps failing during `./gradlew build`, `> Task :integ-test:jarHell FAILED`

```
| sun.boot.class.path: null
| java.home: /Library/Java/JavaVirtualMachines/jdk-14.0.2.jdk/Contents/Home
| examining directory: /Users/szhongna/Desktop/Projects/OpenSearch/sql/integ-test/build/classes/java/test
| examining directory: /Users/szhongna/Desktop/Projects/OpenSearch/sql/integ-test/build/resources/test
| examining jar: /Users/szhongna/Desktop/Projects/OpenSearch/sql/sql-jdbc/build/libs/opensearch-sql-jdbc-2.0.0.0.jar
| examining jar: /Users/szhongna/Desktop/Projects/OpenSearch/sql/sql-jdbc/build/libs/opensearch-sql-jdbc-1.2.0.0.jar
| Exception in thread "main" java.lang.IllegalStateException: jar hell!
| class: org.opensearch.jdbc.ResultSetImpl
| jar1: /Users/szhongna/Desktop/Projects/OpenSearch/sql/sql-jdbc/build/libs/opensearch-sql-jdbc-2.0.0.0.jar
| jar2: /Users/szhongna/Desktop/Projects/OpenSearch/sql/sql-jdbc/build/libs/opensearch-sql-jdbc-1.2.0.0.jar
|       at org.opensearch.bootstrap.JarHell.checkClass(JarHell.java:317)
|       at org.opensearch.bootstrap.JarHell.checkJarHell(JarHell.java:212)
|       at org.opensearch.bootstrap.JarHell.checkJarHell(JarHell.java:99)
|       at org.opensearch.bootstrap.JarHell.main(JarHell.java:83)

```

It doesn't related to my change. Any idea how can I resolve this? @penghuo @joshuali925

zhongnansu commented:
`doctest` and `gradle test` `gradle integTest` can pass

penghuo commented:
> build/libs/opensearch-sql-jdbc-1.2.0.0.jar

clean you local build folder will help.

zhongnansu commented:
@penghuo Hi Peng, I have moved the req/response model from common-utils to sql repo, and added unit test. CI now can pass. Could you please take a look? Thanks

penghuo commented on ppl/build.gradle:
```diff
@@ -59,6 +59,7 @@ dependencies {
     testImplementation group: 'org.hamcrest', name: 'hamcrest-library', version: '2.1'
     testImplementation group: 'org.mockito', name: 'mockito-core', version: '3.12.4'
 
+    api group: 'org.opensearch', name: 'opensearch', version: "${opensearch_version}"
```
PPL module should not depend on org.opensearch. I think there are 2 options.  (1) Move PPLQueryRequest to plugin module or (2) Create TransportPPLQueryRequest if required

zhongnansu replied:
1. Why ppl module shouldn't depend on org.opensearch?
2. option 1 will introduce circular dependency, due to 
   1. `PPLService`(in ppl project) uses `PPLQueryRequest`(in plugin project)
   2. `createPPLService`(in plugin project) uses `PPLService`(in ppl project)
3. I will probably go for option 3, but I still don't think that's a good option. Because we are maintaining duplicate code. The fields of `TransportPPLQueryRequest` and `PPLQueryRequest` will be mostly the same. Therefore, I still think if allowed, we should add opensearch as dependency of ppl module

penghuo replied:
We try to make PPL/SQL as a query engine which not bind to OpenSearch. OpenSearch is just on storage, we working on add others, e.g. Prometheus. 
* TransportPPLQueryRequest is used between Rest layer and Transport layer which align with opensearch framework.
* PPLQueryRequest is the interface required by PPLService which is not bind to opensearch.

For example, we can run PPL as a library and use memory as storage. In this case, all the Rest and Transport related stuff is exclude, we just create a PPLService.

zhongnansu replied:
Got it, that makes sense to me. I'll update the PR with specific `TransportPPLQueryRequest` model

penghuo reviewed (approved):
Thanks for the change!
